text/0000-impl-specialization.md

```diff
@@ -0,0 +1,1652 @@
+
+## Partial impls
+
+An interesting consequence of specialization is that impls need not (and in fact
+sometimes *cannot*) provide all of the items that a trait specifies. Of course,
+this is already the case with defaulted items in a trait -- but as we'll see,
+that mechanism can be seen as just a way of using specialization.
+
+Let's start with a simple example:
+
+```rust
+trait MyTrait {
+    fn foo(&self);
+    fn bar(&self);
+}
+
+impl<T: Clone> MyTrait for T {
+    default fn foo(&self) { ... }
+    default fn bar(&self) { ... }
+}
+
+impl MyTrait for String {
+    fn bar(&self) { ... }
+}
+```
+
+Here, we're acknowledging that the blanket impl has already provided definitions
+for both methods, so the impl for `String` can opt to just re-use the earlier
+definition of `foo`. This is one reason for the choice of the keyword `default`.
+Viewed this way, items defined in a specialized impl are optional overrides of
+those in overlapping blanket impls.
+
+And, in fact, if we'd written the blanket impl differently, we could *force* the
+`String` impl to leave off `foo`:
+
+```rust
+impl<T: Clone> MyTrait for T {
+    // now `foo` is "final"
+    fn foo(&self) { ... }
+
+    default fn bar(&self) { ... }
+}
+```
+
+Being able to leave off items that are covered by blanket impls means that
+specialization is close to providing a finer-grained version of defaulted items
+in traits -- one in which the defaults can become ever more refined as more is
+known about the input types to the traits (as described in the Motivation
+section). But to fully realize this goal, we need one other ingredient: the
+ability for the *blanket* impl itself to leave off some items. We do this by
+using the `partial` keyword:
+
+```rust
+trait Add<Rhs=Self> {
+    type Output;
+    fn add(self, rhs: Rhs) -> Self::Output;
+    fn add_assign(&mut self, Rhs);
+}
+
+partial impl<T: Clone, Rhs> Add<Rhs> for T {
+    fn add_assign(&mut self, rhs: R) {
+        let tmp = self.clone() + rhs;
+        *self = tmp;
+    }
+}
+```
+
+A subsequent overlapping impl of `Add` where `Self: Clone` can choose to leave
+off `add_assign`, "inheriting" it from the partial impl above.
+
+A key point here is that, as the keyword suggests, a `partial` impl may be
+incomplete: from the above code, you *cannot* assume that `T: Add<T>` for any
+`T: Clone`, because no such complete impl has been provided.
+
+With partial impls, defaulted items in traits are just sugar for a partial
+blanket impl:
+
+```rust
+trait Iterator {
+    type Item;
+    fn next(&mut self) -> Option<Self::Item>;
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, None)
+    }
+    // ...
+}
+
+// desugars to:
+
+trait Iterator {
+    type Item;
+    fn next(&mut self) -> Option<Self::Item>;
+    fn size_hint(&self) -> (usize, Option<usize>);
+    // ...
+}
+
+partial impl<T> Iterator for T {
+    default fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, None)
+    }
+    // ...
+}
+```
```

Comment:
This desugaring is questionable, depending on the semantics of `partial`. Currently, defining a trait where every method has a default obviously does not automatically implement that trait for all `T`. You seem to be implying here that a `partial impl` cannot alone ever actually implement methods on a type (even if it implements every method of the trait), and must always be accompanied by an `impl` that isn't `partial` (even if the body of that `impl` is empty). Are these the semantics you have in mind?